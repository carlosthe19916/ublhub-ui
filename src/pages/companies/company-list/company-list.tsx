import React, { useCallback, useEffect } from "react";
import { RouteComponentProps } from "react-router-dom";

import {
  Bullseye,
  Button,
  ButtonVariant,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  PageSection,
  Title,
  ToolbarGroup,
  ToolbarItem,
} from "@patternfly/react-core";
import {
  IActions,
  ICell,
  IExtraData,
  IRowData,
  ISortBy,
  sortable,
} from "@patternfly/react-table";
import { AddCircleOIcon } from "@patternfly/react-icons";

import {
  AppPlaceholder,
  AppTableWithControls,
  ConditionalRender,
  SimplePageSection,
} from "shared/components";
import { useTableControls } from "shared/hooks";

import { formatPath, Paths } from "Paths";
import { Company } from "api/models";

import { Welcome } from "./components/welcome";
import useFetchCompany from "./hooks/useFetchCompany";

const columns: ICell[] = [
  { title: "Name", transforms: [sortable] },
  { title: "Description" },
];

const COMPANY_FIELD = "company";

const getRow = (rowData: IRowData): Company => {
  return rowData[COMPANY_FIELD];
};

const itemsToRow = (items: Company[]) => {
  return items.map((item) => ({
    [COMPANY_FIELD]: item,
    cells: [
      {
        title: item.name,
      },
      {
        title: item.description,
      },
    ],
  }));
};

export interface CompanyListProps extends RouteComponentProps {}

export const CompanyList: React.FC<CompanyListProps> = ({ history }) => {
  const {
    companies,
    isFetching,
    fetchError,
    fetchCount,
    fetchCompanies,
  } = useFetchCompany(true);

  const {
    filterText,
    pagination,
    sortBy,
    handleFilterTextChange,
    handlePaginationChange,
    handleSortChange,
  } = useTableControls();

  const reloadTable = useCallback(
    (
      filterText: string,
      pagination: { page: number; perPage: number },
      sortBy?: ISortBy
    ) => {
      //sortBy.
      fetchCompanies(pagination, undefined, filterText);
    },
    [fetchCompanies]
  );

  useEffect(() => {
    reloadTable(filterText, pagination, sortBy);
  }, [filterText, pagination, sortBy, reloadTable]);

  const actions: IActions = [
    {
      title: "Edit",
      onClick: (
        event: React.MouseEvent,
        rowIndex: number,
        rowData: IRowData,
        extraData: IExtraData
      ) => {
        const row: Company = getRow(rowData);
        history.push(formatPath(Paths.editCompany, { company: row.id }));
      },
    },
  ];

  const newCompany = () => {
    history.push(Paths.newCompany);
  };

  if (fetchCount === 1 && companies?.data.length === 0) {
    return (
      <Bullseye>
        <Welcome onPrimaryAction={newCompany} />
      </Bullseye>
    );
  }

  return (
    <>
      <ConditionalRender
        when={isFetching && !(companies || fetchError)}
        then={<AppPlaceholder />}
      >
        <SimplePageSection
          title="Companies"
          description="Companies allow you to group your documents."
        />
        <PageSection>
          <AppTableWithControls
            count={companies ? companies.meta.count : 0}
            items={companies ? companies.data : []}
            itemsToRow={itemsToRow}
            pagination={pagination}
            sortBy={sortBy}
            handleFilterTextChange={handleFilterTextChange}
            handlePaginationChange={handlePaginationChange}
            handleSortChange={handleSortChange}
            columns={columns}
            actions={actions}
            isLoading={isFetching}
            loadingVariant="skeleton"
            fetchError={fetchError}
            toolbar={
              <ToolbarGroup variant="button-group">
                <ToolbarItem>
                  <Button
                    type="button"
                    variant={ButtonVariant.primary}
                    onClick={newCompany}
                  >
                    New company
                  </Button>
                </ToolbarItem>
              </ToolbarGroup>
            }
            filtersApplied={filterText.trim().length > 0}
            noDataState={
              <EmptyState variant={EmptyStateVariant.small}>
                <EmptyStateIcon icon={AddCircleOIcon} />
                <Title headingLevel="h2" size="lg">
                  No companies available
                </Title>
                <EmptyStateBody>
                  Add a company by clicking on <strong>New company</strong>.
                </EmptyStateBody>
              </EmptyState>
            }
          />
        </PageSection>
      </ConditionalRender>
    </>
  );
};
