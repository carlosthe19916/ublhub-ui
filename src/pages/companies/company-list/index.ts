export { CompanyList as default } from "./company-list";
