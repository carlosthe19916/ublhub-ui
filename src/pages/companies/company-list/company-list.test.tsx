import React from "react";
import { shallow } from "enzyme";
import { CompanyList } from "./company-list";
import { RouteComponentProps } from "react-router-dom";

describe("CompanyList", () => {
  it("Renders without crashing", () => {
    // const route: RouteComponentProps = {
    //   match: {} as any,
    //   location: {} as any,
    //   history: {} as any,
    // };
    // const wrapper = shallow(<CompanyList {...route} />);
    // expect(wrapper).toMatchSnapshot();
  });
});
