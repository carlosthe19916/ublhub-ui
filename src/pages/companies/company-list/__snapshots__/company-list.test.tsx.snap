// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyList Renders without crashing 1`] = `
<Fragment>
  <ConditionalRender
    then={<AppPlaceholder />}
    when={true}
  >
    <SimplePageSection
      description="Companies allow you to group your documents."
      title="Companies"
    />
    <PageSection>
      <AppTableWithControls
        actions={
          Array [
            Object {
              "onClick": [Function],
              "title": "Edit",
            },
          ]
        }
        columns={
          Array [
            Object {
              "title": "Name",
              "transforms": Array [
                [Function],
              ],
            },
            Object {
              "title": "Description",
            },
          ]
        }
        count={0}
        filtersApplied={false}
        handleFilterTextChange={[Function]}
        handlePaginationChange={[Function]}
        handleSortChange={[Function]}
        isLoading={true}
        items={Array []}
        itemsToRow={[Function]}
        loadingVariant="skeleton"
        noDataState={
          <EmptyState
            variant="small"
          >
            <EmptyStateIcon
              icon={[Function]}
            />
            <Title
              headingLevel="h2"
              size="lg"
            >
              No companies available
            </Title>
            <EmptyStateBody>
              Add a company by clicking on 
              <strong>
                New company
              </strong>
              .
            </EmptyStateBody>
          </EmptyState>
        }
        pagination={
          Object {
            "page": 1,
            "perPage": 10,
          }
        }
        toolbar={
          <ForwardRef
            variant="button-group"
          >
            <ToolbarItem>
              <Button
                aria-label="new-company"
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                New company
              </Button>
            </ToolbarItem>
          </ForwardRef>
        }
      />
    </PageSection>
  </ConditionalRender>
</Fragment>
`;
